### ISSUE TITLE
xit() for ignored/pending test

### ISSUE THREAD

**Issue by greghaskins** [greghaskins/spectrum#5]
```java
xit("An ignored test", () -> {
    // will not run
});
```

**Comment by algonzalez** [comment]
I would find it useful, but I use wildcard static imports in certain contexts, like testing, and so can skip the extra static import for `xit`. I know it is frowned upon in certain circles, but I do the following for imports:
```java
import org.junit.runner.RunWith;
import com.greghaskins.spectrum.Spectrum;
import static com.greghaskins.spectrum.Spectrum.*;
// if using hamcrest:
import static org.hamcrest.MatcherAssert.assertThat;
import static org.hamcrest.Matchers.*;
// otherwise:
import static org.junit.Assert.*;
```
I've forked Spectrum and implemented `xit()` and an overloaded version of `it()` that takes an enum (`TestType.IGNORE` and eventually `TestType.FOCUS). The latter does require an extra import for the `TestType` enum.  

I'm testing my changes and looking at xdescribe as well. I'll let you know what progress I make and maybe submit a pull request if you want.

**Comment by algonzalez** [comment]
I wonder, should a null block cause the spec/text to be ignored? I believe this is the way that Jasmine handles it.

**Comment by greghaskins** [comment]
Nice, @algonzalez I look forward to seeing your PR. Might just be the kick in the pants I need to tie up some loose ends on this project.

By a null block, do you mean something like this?

```java
it("is a pending test without a block, and JUnit flags it as Ignored");
```
If so, I like the idea. Not sure if that's how Jasmine works, but it's how it *should* work. 

I was also wondering if something like this might be viable (even if too clever for its own good):

```java
it("is a normal test", () -> {
    // ...
});

it("is an ignored test", (IGNORE) -> {
  // For some Enum or Object parameter, this overload would be treated as an ignored spec
});

**Comment by algonzalez** [comment]
I just want to see if I'm on an acceptable track before submitting a PR. I also need to add some documentation. Would you take a look at this branch of my fork: https://github.com/algonzalez/spectrum/tree/ImplementIgnore

**New:**

- New SpecSection and Taggable interfaces implemented by Context and Test. 
  - Allows method chaining: describe("A Spec", ()->{}).tagWith("SLOW").ignoreWhen((x)->x.hasTag("SLOW"))
  - Provides: ignore(), ignoreWhen(conditionPredicate), onlyWhen(conditionPredicate), tagWith(tagString(s)) and hasTag(tagString) methods.
- describe() and it() without a block will be ignored and marked as pending.
- xdescribe() and xit() will ignore that section of the spec.

**Questions:**

- Use the Predicate interface from Java 8 or our own nested in Spectrum?
- Rename ignoreWhen to skipWhen? 
  - Pro: fewer characters and I think it reads better.
  - Con: JUnit says that test is ignored and some other libraries seem to use the term also
- Should EmptyBlock be a singleton? It will never have any behavior, so can always use the same instance.

**ToDo:**

- discuss questions
- add documentation comments
- update README.md

**Comment by greghaskins** [comment]
Nice work carrying this forward. I checked out your branch and have a few thoughts.

I like the idea of method chaining as a workaround to needing additional imports; I hadn't thought of that. Having `.ignore()` *after* the test block hides that information a bit, though. It should be obvious when a test is ignored, and having that at the top is probably better than at the bottom.

It's pretty hard (for me at least) to visually spot the `.ignore()` in the block below, for example:

```java
it("does something", () -> {
    String expectedResult = "foo";
    assertThat(actualResult.value(), is(expectedResult));
}).ignore();
```
FWIW, `xit()` is not super obvious either, but at least it's easy to spot when scanning a bunch of `it()`s. I also like the idea of keeping our API mostly in line with Jasmine and friends when possible to make it easier for people familiar with those tools.

I can see why ignoring tests conditionally might be desirable, but when things get fancy with conditionals in test code, it's usually a smell of some sort. When that happens, the tests are trying to tell you something about the structure of your code. I looked into how Jasmine and RSpec do this, and both tools have avoided adding much functionality to support that. It seems like the least terrible way (if it needs to be done at all) is to surround tests with conditional statements to make the intention as clear as possible.

```java
if (isRunningOnLinux) {

	it("only runs this on Linux", () -> {
	
	});

}
```

I guess the theme I'm coming around to (despite my own desire to experiment) is for Spectrum to be maximally useful and applicable, perhaps it's best to stick closely to the patterns established by tools like Jasmine, just translated over to Java. That gives us the benefit of consistency with a plethora of frameworks that all act pretty much the same (Jasmine/Mocha, RSpec, Ginkgo, Quick/Kiwi/Cedar, Pho, cspec, specter, etc.). Those patterns have served me well when I code on other platforms, so why reinvent the wheel?

Suggestion: For now, let's stick with:

- `xdescribe("...", () -> {});`
- `xit("...", () -> {});`
- `describe("..."); // without a block`
- `it("..."); // without a block` 

for ignoring tests. (I think you're right that "ignore" is probably the better term. "Skip" somehow feels weird when we're running on top of JUnit.)

---

**Separately: Tags.** The idea of tagging tests is an interesting one, and there is some precedent in RSpec and Cucumber – the implementation is worth discussing in a separate issue from this one (I opened #9).
